### PR TITLE
Change PrefetchHelper to template class

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Refactor PrefetchHelper, QueryMetadataJob, DownloadItemsJob
to work with different roots types.
With PrefetchHelper we can start query data handles for
different types. User should only specify such functions as
'query' - to request metadata needed to download data,
'filter' - to manage results received after query,
'download' - download data.

Relates-To: OLPEDGE-1900

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  size_t total_download_task_count_{0};
+
+  size_t requests_succeeded_{0};
+  size_t requests_failed_{0};
+
+  client::NetworkStatistics accumulated_statistics_;
+
+  PrefetchResult prefetch_result_;
+
+  std::mutex mutex_;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h:141]: Can you please remove the newlines between the members. There is no reason to separate the members in groups.

File: olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  client::NetworkStatistics accumulated_statistics_;
+
+  boost::optional<client::ApiError> query_error_;
+
+  std::shared_ptr<DownloadItemsJob<ItemType, PrefetchResult>> download_job_;
+  std::shared_ptr<thread::TaskScheduler> task_scheduler_;
+  std::shared_ptr<client::PendingRequests> pending_requests_;
+  client::CancellationContext execution_context_;
+
+  std::mutex mutex_;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/QueryMetadataJob.h:163]: Same here, no reason to separate the members in groups.

File: olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+using ExtendedDataResponse = ExtendedApiResponse<model::Data, client::ApiError,
+                                                 client::NetworkStatistics>;
+
+
+// Prototype of function used to download data using data handle.
+using DownloadFunc = std::function<ExtendedDataResponse(
+    std::string, client::CancellationContext)>;
+
+template <typename PrefetchItemsResult>
+using PrefetchItemsResponseCallback = Callback<PrefetchItemsResult>;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h:43]: There are three similar defines, but none seem to be used. Is it needed?

File: olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+class TokenHelper {
+ public:
+  using VectorOfTokens = std::vector<olp::client::CancellationToken>;
+
+  static olp::client::CancellationToken CreateToken(VectorOfTokens tokens) {
```
[the remark this answers is not in the document]
Comment (andescu) [reply olp-cpp-sdk-dataservice-read/src/DownloadItemsJob.h:49]: What I mean is that a simple static function should do the same and no need for an extra class with a static method.

**Review by andescu — APPROVED** [review]

**Review by kzvieriev — APPROVED** [review]

**Pull request merged**